workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceBindings.java

```diff
@@ -252,58 +257,172 @@ private void put(Context context) throws NoSuchWorkspaceException, IOException {
-    final var pathInfo = PathInformation.of(context);
+  private void post(Context context) {
+
+    final String helpText = """
+    Expected JSON body with one of the following formats:
+
+    To move a file:
+    {
+      "moveTo": "<destination-path>",
+      "toWorkspace": <new-workspace-id>, (optional)
+    }
 
-    try (final var bodyReader = Json.createReader(new StringReader(context.body()))) {
-      final var bodyJson = bodyReader.readObject();
+    To copy a file:
+    {
+      "copyTo": "<destination-path>",
+      "toWorkspace": <new-workspace-id>, (optional)
+    }
+    """;
 
-      // Parse what the post request is for
-      final var moveRq = bodyJson.containsKey("moveTo");
+    try (JsonReader bodyReader = Json.createReader(new StringReader(context.body()))) {
+      JsonObject bodyJson = bodyReader.readObject();
 
-      // Perform the request
-      if (moveRq) {
-        final var destination = Path.of(bodyJson.getString("moveTo"));
+      if (bodyJson.containsKey("moveTo")) {
+        handleMove(context, bodyJson);
+      } else if (bodyJson.containsKey("copyTo")) {
+        handleCopy(context, bodyJson);
+      } else {
+        context.status(400).result("Invalid request. Must include either 'moveTo' or 'copyTo' key.\n\n" + helpText);
+      }
+
+      context.status(200).result("Success");
+
+    } catch (JsonException e) {
+      // Malformed JSON in request body
+      context.status(400).result("Malformed JSON: " + e.getMessage() + "\n\n" + helpText);
+
+    } catch (IllegalArgumentException e) {
+      // Logical errors or unsupported operations
+      context.status(400).result("Invalid request: " + e.getMessage() + "\n\n" + helpText);
+
+    } catch (NoSuchWorkspaceException e) {
+      // Workspace not found
+      context.status(404).result("Workspace not found: " + e.getMessage());
+
+    } catch (IOException | SQLException e) {
+      // Internal server error
+      e.printStackTrace();
+      context.status(500).result("Internal server error while processing the request.");
+
+    } catch (Exception e) {
+      // Catch-all for unexpected issues
+      e.printStackTrace();
+      context.status(500).result("Unexpected error: " + (e.getMessage() != null ? e.getMessage() : "Unknown error\n\n" + helpText));
+    }
+  }
+
+  private record CopyMoveValid(int status, String message){};
+
+  private CopyMoveValid isCopyOrMoveValid(int sourceWorkspace, Path sourceFile, int targetWorkspace, Path targetFile)
+  throws NoSuchWorkspaceException {
+    // Reject if source does not exist
+    if (!workspaceService.checkFileExists(sourceWorkspace, sourceFile)) {
+      return new CopyMoveValid(404, sourceFile + " does not exist in the source workspace.");
+    }
+
+    // Reject if target workspace does not exist
+    if (!workspaceService.checkFileExists(targetWorkspace, Path.of("/"))) {
+      return new CopyMoveValid(404, "Target workspace with ID "+targetWorkspace+" does not exist.");
+    }
```

Comment:
I've mixed feelings on keeping these checks, as they're already checked by the actual move/copy procedure, and quite quickly too (especially checking if the other workspace exists).

Reply:
Notably this doesn't seem to be the case in the `moveDirectory` function, which only checks for root directory stuff, but not existence/replacement of existing directories. I would be in favor of only checking in one place, if we can audit & guarantee they're really checked everywhere they need to be. But like you said, nonblocking.